File: services/webmail/internal/domain/large_files.go
package domain

import (
	"errors"
	"time"
)

// Ficheros grandes por enlace (docs/adr/0014). Los guarda mail-files en la base de la empresa y en el
// almacen de objetos; el webmail solo los lleva con la empresa y el buzon de la sesion. Topes, cuotas,
// caducidad y descargas son de mail-files y llegan con cada listado.

var (
	// ErrLargeFilesDisabled: este webmail no tiene mail-files configurado.
	ErrLargeFilesDisabled = errors.New("el envio de ficheros grandes no esta disponible")
	// ErrLargeFileTooLarge: la subida supera lo que el webmail deja pasar hacia mail-files.
	ErrLargeFileTooLarge = errors.New("el fichero supera el tamano maximo")
)

// LargeFileLimits son los topes que aplica mail-files.
type LargeFileLimits struct {
	MaxFileBytes        int64
	DefaultExpiryDays   int
	MaxExpiryDays       int
	DefaultMaxDownloads int
	MaxDownloads        int
	MailboxQuotaBytes   int64
	TenantQuotaBytes    int64
	MaxActivePerMailbox int
}

// LargeFileUsage es lo que ocupan los enlaces vigentes del buzon y de su empresa.
type LargeFileUsage struct {
	MailboxBytes  int64
	MailboxActive int
	TenantBytes   int64
}

// LargeFile es un fichero compartido por el buzon. URL solo viene mientras el enlace sirve.
type LargeFile struct {
	ID                 string
	Name               string
	SizeBytes          int64
	SHA256             string
	URL                string
	State              string
	ExpiresAt          time.Time
	MaxDownloads       int
	Downloads          int
	RemainingDownloads int
	CreatedAt          time.Time
	LastDownloadAt     *time.Time
	RevokedAt          *time.Time
}

// LargeFileListing es lo que ve el remitente: si la funcion esta disponible, sus enlaces, su uso y los
// topes.
type LargeFileListing struct {
	Enabled bool
	Items   []LargeFile
	Usage   LargeFileUsage
	Limits  LargeFileLimits
}

// LargeFileOptions son las elecciones del remitente; 0 deja el valor por defecto de mail-files.
type LargeFileOptions struct {
	ExpiresInDays int
	MaxDownloads  int
}

File: web/src/api/largeFiles.ts
import { endpoints } from './endpoints';
import { webmailRequest } from './webmail';

/*
 * Ficheros grandes por enlace (services/mail-files a traves de services/webmail, docs/adr/0014). El
 * fichero se sube con la sesion del buzon, mail-files lo analiza con ClamAV y lo guarda, y el correo
 * lleva el enlace. Topes, cuotas, caducidad y descargas los sirve el servicio en cada listado.
 */

/** Estado que ve el remitente (domain.State de mail-files). */
export const LARGE_FILE_STATES = [
  'uploading',
  'active',
  'expired',
  'exhausted',
  'revoked',
  'failed',
] as const;
export type LargeFileState = (typeof LARGE_FILE_STATES)[number];

export interface LargeFileLimits {
  max_file_bytes: number;
  default_expiry_days: number;
  max_expiry_days: number;
  default_max_downloads: number;
  max_downloads: number;
  mailbox_quota_bytes: number;
  tenant_quota_bytes: number;
  max_active_per_mailbox: number;
}

export interface LargeFileUsage {
  mailbox_bytes: number;
  mailbox_active: number;
  tenant_bytes: number;
}

export interface LargeFile {
  id: string;
  name: string;
  size_bytes: number;
  sha256: string;
  /** Solo mientras el enlace sirve (state active). */
  url: string;
  state: LargeFileState;
  expires_at: string;
  max_downloads: number;
  downloads: number;
  remaining_downloads: number;
  created_at: string;
  last_download_at: string | null;
  revoked_at: string | null;
}

/** GET /webmail/large-files. enabled false: la funcion no esta disponible y no se ofrece. */
export interface LargeFileListing {
  enabled: boolean;
  items: LargeFile[];
  usage: LargeFileUsage;
  limits: LargeFileLimits;
}

export interface LargeFileOptions {
  expiresInDays?: number;
  maxDownloads?: number;
}

const wm = endpoints.webmail;

export const largeFilesApi = {
  list: (signal?: AbortSignal) =>
    webmailRequest<LargeFileListing>('GET', wm.largeFiles, { signal }),
  /** Sube el fichero en multipart; el servicio responde cuando ClamAV lo dio por limpio y esta guardado. */
  upload: (file: File, options: LargeFileOptions = {}) => {
    const form = new FormData();
    form.append('file', file, file.name);
    return webmailRequest<LargeFile>('POST', wm.largeFiles, {
      form,
      params: { expires_in_days: options.expiresInDays, max_downloads: options.maxDownloads },
    });
  },
  revoke: (id: string) => webmailRequest<LargeFile>('DELETE', wm.largeFile(id)),
};
